mat64/gsvd.go

```diff
@@ -0,0 +1,296 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+// Based on the SingularValueDecomposition class from Jama 1.0.3.
+
+package mat64
+
+import (
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/floats"
+	"github.com/gonum/lapack"
+	"github.com/gonum/lapack/lapack64"
+	"github.com/gonum/matrix"
+)
+
+// GSVD is a type for creating and using the Generalized Singular Value Decomposition
+// (GSVD) of a matrix.
+type GSVD struct {
+	kind matrix.GSVDKind
+
+	m, p, n, k, l int
+	c, s          []float64
+	a, b, u, v, q blas64.General
+
+	work  []float64
+	iwork []int
+}
+
+// Factorize computes the generalized singular value decomposition (GSVD) of the input
+// the m×n matrix A and the p×n matrix B. The singular values of A and B are computed
+// in all cases, while the singular vectors are optionally computed depending on the
+// input kind.
+//
+// The full singular value decomposition (kind == GSVDU|GSVDV|GSVDQ) deconstructs A and B as
+//  A = U * Σ₁ * [ 0 R ] * Q^T
+//
+//  B = V * Σ₂ * [ 0 R ] * Q^T
+// where Σ₁ and Σ₂ are m×(k+l) and p×(k+l) diagonal matrices of singular values, and
```

Comment:
Could you add though, the " Let K+L = the effective numerical rank of the matrix (A**T,B**T)**T" into the main description. There's no comment on what k and l are in the factorize documentation.